File: aug2026/abc469/abc469_f.py

import sys
from itertools import combinations

sys.setrecursionlimit(10**7)


def read_tokens():
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            tokens = f.read().split()
    else:
        tokens = sys.stdin.buffer.read().split()
    return [t.decode() if isinstance(t, bytes) else t for t in tokens]


class unionfind:
    def __init__(self, values):
        self.parent = {value: value for value in values}
        self.size = {value: 1 for value in values}
        self.components = len(self.parent)

    def root(self, value):
        if self.parent[value] != value:
            self.parent[value] = self.root(self.parent[value])
        return self.parent[value]

    def union(self, first, second):
        first_root = self.root(first)
        second_root = self.root(second)
        if first_root == second_root:
            return

        if self.size[first_root] < self.size[second_root]:
            first_root, second_root = second_root, first_root

        self.parent[second_root] = first_root
        self.size[first_root] += self.size[second_root]
        self.components -= 1

    def numclusters(self):
        return self.components


data = read_tokens()
if not data:
    raise SystemExit

n = int(data[0])
a = [int(data[i]) for i in range(1, n + 1)]
max_a = max(a)


def solve():
    index_of = [-1] * (max_a + 1)
    for index, value in enumerate(a):
        index_of[value] = index

    ans = 0
    clusters = unionfind(range(n))

    for i in range(max_a, 0, -1):
        dividing_node = -1
        for multiple in range(i, max_a + 1, i):
            a_idx = index_of[multiple]
            if a_idx == -1:
                continue
            if dividing_node == -1:
                dividing_node = a_idx
            elif clusters.root(dividing_node) != clusters.root(a_idx):
                clusters.union(dividing_node, a_idx)
                ans += i

                if clusters.numclusters() == 1:
                    return ans

    return ans


print(solve())

File: aug2026/abc469/abc469_f.cpp
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

class UnionFind {
public:
    explicit UnionFind(int n) : parent(n), component_size(n, 1), components(n) {
        iota(parent.begin(), parent.end(), 0);
    }

    int root(int value) {
        if (parent[value] != value) {
            parent[value] = root(parent[value]);
        }
        return parent[value];
    }

    bool unite(int first, int second) {
        int first_root = root(first);
        int second_root = root(second);
        if (first_root == second_root) {
            return false;
        }

        if (component_size[first_root] < component_size[second_root]) {
            swap(first_root, second_root);
        }

        parent[second_root] = first_root;
        component_size[first_root] += component_size[second_root];
        --components;
        return true;
    }

    int num_components() const {
        return components;
    }

private:
    vector<int> parent;
    vector<int> component_size;
    int components;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<int> a(n);
    int max_a = 0;
    for (int& value : a) {
        cin >> value;
        max_a = max(max_a, value);
    }

    UnionFind clusters(n);
    long long answer = 0;

    vector<int> index_of(max_a + 1, -1);
    for (int index = 0; index < n; ++index) {
        index_of[a[index]] = index;
    }

    for (int divisor = max_a; divisor >= 1; --divisor) {
        int representative = -1;
        for (int multiple = divisor; multiple <= max_a; multiple += divisor) {
            int index = index_of[multiple];
            if (index == -1) {
                continue;
            }

            if (representative == -1) {
                representative = index;
            } else if (clusters.unite(representative, index)) {
                answer += divisor;
                if (clusters.num_components() == 1) {
                    cout << answer << '\n';
                    return 0;
                }
            }
        }
    }

    cout << answer << '\n';
    return 0;
}
